src/GasEstimator.ts

```diff
@@ -0,0 +1,230 @@
+ * @param request request that contains the relayRequest/deployRequest and metadata
+ * @param relayWorkerAddress address of the relayWorker that will execute the transaction
+ * @returns gas estimation from the relayTransaction
+ */
+export const estimateRelayMaxPossibleGas = async (
+    contractInteractor: ContractInteractor,
+    request: RelayTransactionRequest | DeployTransactionRequest,
+    relayWorkerAddress: string
+): Promise<BigNumber> => {
+    const { relayRequest, metadata } = request;
+
+    let estimation: BigNumber;
+
+    const tokenEstimation = await estimateMaxPossibleGasTokenTransfer(
+        contractInteractor,
+        relayRequest
+    );
+
+    if (metadata.signature > '0x0') {
+        estimation = await standardMaxPossibleGasEstimation(
+            contractInteractor,
+            request,
+            relayWorkerAddress,
+            tokenEstimation
+        );
+    } else {
+        estimation = await linearFitMaxPossibleGasEstimation(
+            contractInteractor,
+            relayRequest,
+            tokenEstimation
+        );
+    }
+
+    return estimation;
+};
+
+/**
+ * Check if the relay request is a smart wallet deployment request. Transaction request does not contain the `index` property of the former.
+ * @param request to be relayed. May be either transaction, or smart-wallet-deployment request.
+ * @returns `true` if given request is a smart wallet deployment request; `false` if a transaction request.
+ */
+const isDeployRequest = (
+    request: DeployRequestStruct | ForwardRequest
+): boolean => {
+    return 'index' in request;
+};
+
+/**
+ * Estimates the max possible gas consumed by relaying a transaction using a standard estimation
+ * @param contractInteractor object containing the contractInteractor that is used to interact with the blockchain
+ * @param request request that contains the relayRequest/deployRequest and metadata
+ * @param relayWorkerAddress address of the relayWorker that will execute the transaction
+ * @param tokenEstimation gas consumed by the token transfer
+ * @returns gas estimation from the relayTransaction
+ */
+export const standardMaxPossibleGasEstimation = async (
+    contractInteractor: ContractInteractor,
+    {
+        relayRequest,
+        metadata
+    }: RelayTransactionRequest | DeployTransactionRequest,
+    relayWorkerAddress: string,
+    tokenEstimation: BigNumber
+): Promise<BigNumber> => {
+    const { request, relayData } = relayRequest;
+
+    let methodToEstimate;
+    if (isDeployRequest(request)) {
+        methodToEstimate =
+            contractInteractor.relayHubInstance.contract.methods.deployCall(
+                relayRequest as DeployRequest,
+                metadata.signature
+            );
+    } else {
+        methodToEstimate =
+            contractInteractor.relayHubInstance.contract.methods.relayCall(
+                relayRequest as RelayRequest,
+                metadata.signature
+            );
+    }
```

Comment:
I know @antomor is not a big fan of this approach, but I'd still suggest it as it makes it so much simpler.
```suggestion
    const callType = isDeployRequest(request) ? 'deployCall' : 'relayCall';
    const methodToEstimate = contractInteractor.relayHubInstance.contract.methods[callType](
        relayRequest,
        metadata.signature
    );
```

Reply:
updated

Reply:
Yes, not a fan of this approach, but considering that `contract` is of type `any`, not a big issue 😓